consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,177 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows this node to identify itself to the peer as a validator and vice versa
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	errCh := make(chan error, 4)
+	isValidatorCh := make(chan bool, 1)
+
+	go func() {
+		validatorProofMessage, err := sb.generateValidatorProofMessage(peer)
+		errCh <- err
+		if err != nil {
+			return
+		}
+		msgBytes, err := validatorProofMessage.Payload()
+		errCh <- err
+		if err != nil {
+			return
+		}
+		errCh <- peer.Send(istanbulValidatorProofMsg, msgBytes)
+	}()
+	go func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		errCh <- err
+		isValidatorCh <- isValidator
+	}()
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	// We write to errCh four times unless if the timeout is triggered or an error isn't nil
+	for i := 0; i < 4; i++ {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return false, err
+			}
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		}
+	}
+	return <-isValidatorCh, nil
+}
+
+// generateValidatorProofMessage will create a message that contains a proof
+// to the peer during the handshake that this node is a validator, which is just
+// a versioned enode message.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node
+// and an empty message is created.
+// If this node is a proxy, it will use the most recent versioned enode message this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
+	shouldSend, err := sb.shouldSendValidatorProof(peer)
+	if err != nil {
+		return nil, err
+	}
+	if shouldSend {
+		msg, err := sb.getSelfVersionedEnodeMsg(getCurrentAnnounceVersion())
+		if err != nil {
+			return nil, err
+		}
+		if msg != nil {
+			return msg, nil
+		}
+	}
+	// Even if we decide not to identify ourselves,
+	// send an empty message to complete the handshake
+	return &istanbul.Message{}, nil
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
```

Comment:
⚠️ This function can potentially block for multiple seconds. If that happens on the path of a handshake, the connection could timeout. ⚠️

Reply:
I don't think there's a straightforward way around this.. I'm going to up the handshake timeout from 5 to 10 seconds, but if you have any ideas I'd love to hear them.

Reply:
I would recommend leaving the handshake timeout as 5 seconds, and refactor this code a bit to allow for retrieving the cached value without triggering an update (or at least not a synchronous update). It'd be better to use stale data here than timeout the handshake.

Reply:
Thought about this and I agree, I pushed some code that uses a previously cached value and asynchronously updates